container/kvm/kvm.go

```diff
@@ -213,14 +215,25 @@ func (manager *containerManager) CreateContainer(
 		return nil, nil, errors.Annotate(err, "failed to parse hardware")
 	}
 
-	callback(status.Provisioning, "Creating container; it might take some time", nil)
+	_ = callback(status.Provisioning, "Creating container; it might take some time", nil)
 	logger.Tracef("create the container, constraints: %v", cons)
+
+	// Lock around finding an image.
+	// The provisioner works concurrently to create containers.
+	// If an image needs to be copied from a remote, we don't many goroutines
+	// attempting to do it at once.
+	manager.imageMutex.Lock()
+	err = kvmContainer.EnsureCachedImage(startParams)
```

Comment:
shouldn't that lock better be "inside" the function to ensure if it is ever called from another place that they still use the same lock? If you prefer it here then the function might want a way to check if the right lock is held maybe?

Reply:
The mutex itself needs to be a member of the (singular) manager rather than the container object, as that's the level at which we have to gate access.

We could either instantiate the container with the mutex passed in, or call the method passing the mutex, but handing around mutex references has a smell about it.

This code has parts that are half a decade old. It's not ideal but I've gone for the "lightest touch". It has also never been Goroutine safe in that time - I have a follow-up patch that will address this.